### PR TITLE
FreeRTOS_ARP.c : store local addresses only

### PR DESCRIPTION
**Pull request by htibosch**
Description
-----------
**Note:** the name of my branch is incorrect, it should have been "arp_store_local_addresses_only", in stead of referring to "dhcp".

Three changes to FreeRTOS_ARP.c :

1) When `ipconfigARP_STORES_REMOTE_ADDRESSES` is `0`, ARP doesn't want to store any non-local IP-address in the ARP cache.
We made an exception, during the DHCP phase, when `*ipLOCAL_IP_ADDRESS_POINTER` is still zero, any address would be stored:
~~~c
#if ( ipconfigARP_STORES_REMOTE_ADDRESSES == 0 )
    if( ( ( ulIPAddress & xNetworkAddressing.ulNetMask ) == ( ( *ipLOCAL_IP_ADDRESS_POINTER ) & xNetworkAddressing.ulNetMask ) ) ||
        ( *ipLOCAL_IP_ADDRESS_POINTER == 0UL ) )
    {
        /* Store the address. */
    }
~~~

DHCP doesn't need ARP beecause the client uses the IP broadcast address to find the DHCP server. I would like to change the test to:

~~~c
#if ( ipconfigARP_STORES_REMOTE_ADDRESSES == 0 )
    /* Only process the IP address if it is on the local network. */
    if( ( ulIPAddress & xNetworkAddressing.ulNetMask ) == ( ( *ipLOCAL_IP_ADDRESS_POINTER ) & xNetworkAddressing.ulNetMask ) )
    {
        /* Store the address. */
    }
~~~

2) In the function `eARPGetCacheEntry()`, the MAC-address of the LLMNR IP-address was still set 'manually'. That is not necessary since it would have been set a few lines later in:
~~~c
    if( xIsIPv4Multicast( ulAddressToLookup ) != 0 )
    {
        /* Get the lowest 23 bits of the IP-address. */
        vSetMultiCastIPv4MacAddress( ulAddressToLookup, pxMACAddress );
        eReturn = eARPCacheHit;
    }
~~~

3) Resolving the "own" IP-address in eARPGetCacheEntry() : I would like to add these lines:
~~~c
    else if( *ipLOCAL_IP_ADDRESS_POINTER == *pulIPAddress )
    {
        /* The address of this device. May be useful for the loopback device. */
        eReturn = eARPCacheHit;
        memcpy( pxMACAddress->ucBytes, ipLOCAL_MAC_ADDRESS, sizeof( pxMACAddress->ucBytes ) );
    }
~~~
so that a loopback can be used.

4) And finally: the addition of `xARPWaitResolution()`. It can be called by the application to confirm that the MAC-address of a local device is known.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by AniruddhaKanhere** [comment]
@htibosch, can you try to fix the CI checks which are failing so that we can move ahead with the review? If there is some issue with the fix, feel free to ping me.
Thank you.

**Comment by htibosch** [comment]
CI Checks / unittest fails:

Linking `FreeRTOS_TCP_Unit_test` give a few errors because tasks.c is not linked in the unit test project?
These functions are ,missing:

`vTaskSetTimeOutState`
`vTaskDelay`  needs `INCLUDE_vTaskDelay = 1`
`xTaskCheckForTimeOut

Could you check that please? Thanks.

**Review by AniruddhaKanhere — COMMENTED** [review]
See the comments thanks!

File: FreeRTOS_ARP.c [hunk 1]
```diff
@@ -769,7 +761,63 @@ void FreeRTOS_OutputARPRequest( uint32_t ulIPAddress )
...
+ * @brief  Wait for address resolution: look-up the IP-address in the ARP-cache, and if
+ *         needed send an ARP request, and wait for a reply.  This function is useful when
+ *         called before FreeRTOS_sendto().
+ *
+ * @param[in] ulIPAddress: The IP-address to look-up.
+ * @param[in] uxTicksToWait: The maximum number of clock ticks to wait for a reply.
+ *
+ * @return Zero when successful.
+ */
+BaseType_t xARPWaitResolution( uint32_t ulIPAddress,
```
[the remark this answers is not in the document]
Comment (htibosch) [reply FreeRTOS_ARP.c:776]: Ah sorry, I had declared it in FreeRTOS_IP.h, now updated.

**Review by htibosch — COMMENTED** [review]
Updated FreeRTOS_IP.h and replaced 0 with pdFALSE in 2 occasions. Thanks

**Review by AniruddhaKanhere — APPROVED** [review]

**Review by n9wxu — APPROVED** [review]

**Pull request merged**